akka-actor/src/main/scala/akka/actor/ActorSystem.scala

```diff
@@ -783,44 +784,36 @@ private[akka] class ActorSystemImpl(val name: String, applicationConfig: Config,
     printNode(lookupRoot, "")
   }
 
-  final class TerminationCallbacks extends Runnable with Awaitable[Unit] {
-    private val lock = new ReentrantGuard
-    private var callbacks: List[Runnable] = _ //non-volatile since guarded by the lock
-    lock withGuard { callbacks = Nil }
+  final class TerminationCallbacks extends AtomicReference[List[Runnable]](Nil) with Runnable with Awaitable[Unit] {
+    private val done = Promise[Unit]()
 
-    private val latch = new CountDownLatch(1)
-
-    final def add(callback: Runnable): Unit = {
-      latch.getCount match {
-        case 0 ⇒ throw new RejectedExecutionException("Must be called prior to system shutdown.")
-        case _ ⇒ lock withGuard {
-          if (latch.getCount == 0) throw new RejectedExecutionException("Must be called prior to system shutdown.")
-          else callbacks ::= callback
-        }
+    final def add(callback: Runnable): Unit =
+      get match {
+        case null ⇒ throw new RejectedExecutionException("Must be called prior to system shutdown.")
```

Comment:
There is a slight difference in semantics for this exception, where it before threw the exception *after all* callbacks had executed, it now throws the exception if shutdown has started. This can easily be fixed but there does not seem to be a test for that semantics so it was uncertain if it was by design or not.